### PR TITLE
fix: content-type mis-handling for invalid non-essence content-type

### PR DESCRIPTION
**Pull request by climba03003**
@Uzlopak 
We do found a bug throughout the conversation on #4477 
I would fix this case first, but left the leading and trailing space support implementation.
It would be a heavy works for supporting it, since it means we need to do a parsing of input anywhere.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by climba03003** [comment]
I am actually consider we should throw for something like `json;` and suggest to use `RegExp`.
`json;` is means to be match any content-type that end with `json`. Splitting and take a part from it doesn't seem right.

File: lib/contentTypeParser.js [hunk 1]
```diff
@@ -363,6 +363,7 @@ function compareContentType (contentType, parserListItem) {
     // we do essence check
     return contentType.type.indexOf(parserListItem) !== -1
   } else {
+    console.log(parserListItem.type)
```
Comment (Uzlopak) [inline lib/contentTypeParser.js:366]:
```suggestion
```

**Comment by Uzlopak** [comment]
Tbh, i also thought we should throw an error, but in case we have contentType.type === ''. Because it means, that the media type is also potentially broken. 

Like 

```
fastify.addContentTypeParser('اراز', function (request, body, done) {
    t.pass('should be called')
    done(null, body)
  })
```

ParserListItem.type would be now also 'اراز', despite it contains invalid characters.

**Comment by metcoder95** [comment]
In sync with throwing an error for the scenario. Not so sure what is the common denominator for use cases, but it might be a good idea to also reduce its combination to support only the essence of the essence through wildcards fully.

It might come with the cost of performance depending on the wildcards the users submit but will be easier to control.
For regex, something similar as by checking that it contains at least the `/` for separate type/subtype, and recommendations in case of possible ReDoS. But worth it open the chat 🙂

**Review by mcollina — APPROVED** [review]
lgtm

File: lib/contentTypeParser.js [hunk 2]
```diff
@@ -393,10 +393,16 @@ function ParserListItem (contentType) {
   // we pre-calculate all the needed information
   // before content-type comparsion
   const parsed = safeParseContentType(contentType)
-  this.type = parsed.type
+  this.isEssence = contentType.indexOf(';') === -1
+  // we should not allow empty string for parser list item
+  // because it would becomes a match-all handler
```
Comment (Uzlopak) [inline lib/contentTypeParser.js:398]:
```suggestion
  // because it would become a match-all handler
```

File: lib/contentTypeParser.js [hunk 3]
```diff
@@ -393,10 +393,16 @@ function ParserListItem (contentType) {
...
+  if (this.isEssence === false && parsed.type === '') {
+    this.type = contentType.split(';')[0]
```
Comment (Uzlopak) [inline lib/contentTypeParser.js:400]: I thought very long about this. I am not convinced that it will avoid the case that we could replace the wildcard.

E.g. contentType with the value ";" maybe also results being a wildcard

Comment (Uzlopak) [reply lib/contentTypeParser.js:400]: Or am i having a brainfart?

File: test/content-parser.test.js [hunk 4]
```diff
@@ -652,3 +652,69 @@ test('content-type regexp list should be cloned when plugin override', async t =
...
+      'content-type': 'application/json; foo=bar; charset=utf8'
+    },
+    body: ''
+  })
+
+  await fastify.inject({
+    method: 'POST',
+    path: '/',
+    headers: {
+      'content-type': 'image/jpeg'
```
Comment (Eomm) [inline test/content-parser.test.js:716]: It would be interesting to add some edge case also

**Comment by climba03003** [comment]
CI failure seems unrelated. I do not have any MacOS device to debug.
